### PR TITLE
Johanna/azure functions upgrade v4

### PR DESCRIPTION
**Pull request by johanna-skylight**
# DEVOPS PULL REQUEST

## Related Issue

- This PR handles the `@azure/functions` dependency upgrade. Replacing [this dependabot](https://github.com/CDCgov/prime-simplereport/pull/6713) PR. Upgrade to version 4 needed coded changes.

## Changes Proposed

- Replace the `Context` object with ` InvocationContext`.
- Replace `log.error` method with direct `error` method.
- Replace the usage of `function.json` file with the corresponding [app.[type of function]](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-timer?tabs=python-v2%2Cisolated-process%2Cnodejs-v4&pivots=programming-language-typescript) method (E.g.: `app.timer()`, `app.storageQueue()`).
- Update unit testing files. Including adding a mock for the new `app` object.
- Add `main` property to `package.json` so azure knows the [entry point of each function](storageQueue).
- Update host.json with the matching azure function version running.
- Update ts-jest, jest, types/jest so they all use the same typescript version than `@azure/functions`.

## Testing (Code available in dev4)
- All unit testing should pass.
- The processing and reporting of the COVID and Universal pipeline should work as expected.
- The processing and storing of ReportStream errors into our DB should be working as expected.

### CODE REVIEW

**Comment by johanna-skylight** [comment]
Tagging Better Data team as this is a refactor due to an upgrade. (@zdeveloper , @DanielSass , @nathancrtr )

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts [hunk 1]
```diff
@@ -22,10 +22,11 @@ const {
 appInsights.setup();
 const telemetry = appInsights.defaultClient;
 
-const QueueBatchedTestEventPublisher: AzureFunction = async function (
-  context: Context,
+export async function QueueBatchedTestEventPublisher(
+  myTimer: Timer,
```
Comment (fzhao99) [inline ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts:26]: Is the function using this passed-in instance under the hood somehow? Don't see it used anywhere in the file

Comment (johanna-skylight) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts:26]: The function execution is being handled/triggered by azure as before. I needed to add both parameters so the function signature would match the type that `app.timer` is expecting.

**Review by fzhao99 — APPROVED** [review]
left one question but otherwise tested on dev4 and lgtm

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts (hunk 1 above)
Comment (fzhao99) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts:26]: ahh ok, that's what I was guessing. In that case could [we use an underscore to mark that the param isn't being used?](https://stackoverflow.com/questions/27637013/what-is-the-meaning-of-an-underscore-in-javascript-function-parameter#:~:text=A%20single%20underscore%20is%20a,could%20have%20been%20omitted%20entirely.)

Comment (BobanL) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts:26]: I'd agree to add the underscore in front, I checked IntelliJ and they even suggest that if you hover over it.
image

**Review by BobanL — APPROVED** [review]
Everything looks good, just a small nit where I agree with Bob to add an underscore to `myTimer` in both index.ts files.

**Comment by johanna-skylight** [comment]
I have updated the undici dependency. Latest code is in dev4.

**Review by fzhao99 — DISMISSED** [review]

**Comment by nathancrtr** [comment]
@johanna-skylight I submitted some test results and I can see the Azure Functions moving them along the pipeline to RS 🎉 

Screen Shot 2023-10-19 at 5 02 53 PM

I _do_ see associated runs of the `ReportStreamExceptionHandler` when these events are sent. Do you know if this is normal - or how to better inspect what's going on?

Screen Shot 2023-10-19 at 5 13 16 PM

I'll do a little more digging - it's probably fine, but I'll hold off on a formal approval for now.

**Comment by johanna-skylight** [comment]
> @johanna-skylight I submitted some test results and I can see the Azure Functions moving them along the pipeline to RS 🎉
> 
> Screen Shot 2023-10-19 at 5 02 53 PM
> I _do_ see associated runs of the `ReportStreamExceptionHandler` when these events are sent. Do you know if this is normal - or how to better inspect what's going on?
> 
> Screen Shot 2023-10-19 at 5 13 16 PM
> I'll do a little more digging - it's probably fine, but I'll hold off on a formal approval for now.

@nathancrtr yes! This is normal behavior if the response we got back from report stream contained `errors` or `warnings` associated with `event ids`. An example of a common warning in lower envs is `No match found for Test_Kit_Name_ID (test_kit_name_id);...`. Once this warning is received by the function it gets placed in the `error` queue. The ReportStreamExceptionHandler picks it from there and inserts this data into the `ReportStreamResponse` table. This part of the flow should also work as expected with this migration. LMK if you have more questions!.

**Comment by johanna-skylight** [comment]
I have resolved merge conflicts with some dependabot PRs merged. The latest code is available in dev4.

**Review by emyl3 — APPROVED** [review]
tested on dev4 -- LGTM! Thank you for working on this!!

**Review by DanielSass — APPROVED** [review]

**Pull request merged**